File: geotrellis-python/tests/geotrellis_test/spark/io/AttributeStoreSpec.py
from __future__ import absolute_import
from spec import Spec, skip
from nose import tools
from nose.tools import ok_
from tests.geotrellis_test.spark.TestEnvironment import _TestEnvironment
from geotrellis.python.util.utils import JSONFormat
from geotrellis.spark.LayerId import LayerId
from geotrellis.python.spray.json.package_scala import DeserializationException

@tools.nottest
class _AttributeStoreSpec(_TestEnvironment):
    @property
    def attributeStore(self):
        pass

    @tools.istest
    def test_01_should_write_to_an_attribute_store(self):
        """should write to an attribute store"""
        store = self.attributeStore
        store.write(str, LayerId("test1", 1), "test-att1", "test")
        store.write(str, LayerId("test2", 2), "test-att1", "test")
        store.write(str, LayerId("test2", 2), "test-att2", "test")
        store.write(str, LayerId("test3", 3), "test-att1", "test")

    @tools.istest
    def test_02_should_know_that_these_new_IDs_exist(self):
        """should know that these new IDs exist"""
        store = self.attributeStore
        ok_(store.layerExists(LayerId("test1", 1)) == True)
        ok_(store.layerExists(LayerId("test2", 2)) == True)
        ok_(store.layerExists(LayerId("test3", 3)) == True)

    @tools.istest
    def test_03_should_read_layer_ids(self):
        """should read layer ids"""
        store = self.attributeStore
        layerids = sorted(store.layerIds(), key = lambda layerid: layerid.zoom)
        ok_(layerids == [LayerId("test1", 1), LayerId("test2", 2), LayerId("test3", 3)])

    @tools.istest
    def test_04_should_clear_out(self):
        """should clear out the attribute store"""
        store = self.attributeStore
        for layerid in store.layerIds():
            store.delete(layerid)

        length = len(store.layerIds())
        ok_(length == 0)

    @tools.istest
    def test_should_save_and_load_a_random_object(self):
        """should save and load a random object convertable to/from json"""
        layerid = LayerId("test", 3)
        class Foo(object):
            implicits = {"format": lambda: FooFormat()}
            def __init__(self, x, y):
                self.x = x
                self.y = y
            def __eq__(self, other):
                if not isinstance(other, Foo):
                    return False
                return self.x == other.x and self.y == other.y
        class FooFormat(JSONFormat):
            def to_dict(self, obj):
                return {"first": obj.x, "second": obj.y}

            def from_dict(self, dct):
                fields = self.get_fields(dct, "first", "second")
                if not fields:
                    raise DeserializationException("JSON reading failed")
                x, y = fields
                return Foo(x, y)

        foo = Foo(1, "thing")
        store = self.attributeStore
        store.write(Foo, layerid, "foo", foo)
        fromstore = store.read(Foo, layerid, "foo")
        ok_(fromstore == foo)

File: spark/src/test/scala/geotrellis/spark/io/AttributeStoreSpec.scala
/*
 * Copyright 2016 Azavea
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package geotrellis.spark.io

import geotrellis.raster.io._
import geotrellis.raster.histogram._
import geotrellis.layers.{LayerId, AttributeStore}
import geotrellis.spark._
import geotrellis.spark.testkit.testfiles._
import geotrellis.spark.summary._
import geotrellis.spark.testkit._

import org.scalatest._

import spray.json._
import spray.json.DefaultJsonProtocol._


abstract class AttributeStoreSpec
    extends FunSpec
    with Matchers
    with TestEnvironment
    with TestFiles {
  def attributeStore: AttributeStore

  it("should write to an attribute store") {
    attributeStore.write(LayerId("test1", 1), "metadata", "test")
    attributeStore.write(LayerId("test2", 2), "metadata", "test")
    attributeStore.write(LayerId("test2", 2), "metadata", "test")
    attributeStore.write(LayerId("test3", 3), "metadata", "test")
  }

  it("should know that these new IDs exist") {
    attributeStore.layerExists(LayerId("test1", 1)) should be (true)
    attributeStore.layerExists(LayerId("test2", 2)) should be (true)
    attributeStore.layerExists(LayerId("test3", 3)) should be (true)
  }

  it("should read layer ids") {
    attributeStore.layerIds.sortBy(_.zoom).toList should be (List(LayerId("test1", 1), LayerId("test2", 2), LayerId("test3", 3)))
  }

  it("should clear out the attribute store") {
    attributeStore
      .layerIds
      .foreach(attributeStore.delete(_))

    attributeStore
      .layerIds.isEmpty should be (true)
  }

  it("should save and pull out a histogram") {
    val layerId = LayerId("test", 3)
    val histo = DecreasingTestFile.histogramExactInt

    attributeStore.write(layerId, "histogram", histo)

    val loaded = attributeStore.read[Histogram[Int]](layerId, "histogram")
    loaded.mean should be (histo.mean)
  }

  it("should save and load a random RootJsonReadable object") {
    val layerId = LayerId("test", 3)
    case class Foo(x: Int, y: String)

    implicit object FooFormat extends RootJsonFormat[Foo] {
      def write(obj: Foo): JsObject =
        JsObject(
          "the_first_field" -> JsNumber(obj.x),
          "the_second_field" -> JsString(obj.y)
        )

      def read(json: JsValue): Foo =
        json.asJsObject.getFields("the_first_field", "the_second_field") match {
          case Seq(JsNumber(x), JsString(y)) =>
            Foo(x.toInt, y)
          case _ =>
            throw new DeserializationException(s"JSON reading failed.")
        }
    }

    val foo = Foo(1, "thing")

    attributeStore.write(layerId, "foo", foo)
    attributeStore.read[Foo](layerId, "foo") should be (foo)
  }
}
